File: backend/internal/handlers/inventory.go
package handlers

import (
	"database/sql"
	"net/http"
	"time"

	"pos-public/internal/services"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type InventoryHandler struct {
	db              *sql.DB
	notificationSvc *services.NotificationService
}

func NewInventoryHandler(db *sql.DB) *InventoryHandler {
	return &InventoryHandler{
		db:              db,
		notificationSvc: services.NewNotificationService(db),
	}
}

// InventoryItem represents combined product and inventory data
type InventoryItem struct {
	ProductID     uuid.UUID `json:"product_id"`
	ProductName   string    `json:"product_name"`
	CategoryName  string    `json:"category_name"`
	CurrentStock  int       `json:"current_stock"`
	MinStock      int       `json:"min_stock"`
	MaxStock      int       `json:"max_stock"`
	Unit          string    `json:"unit"`
	LastRestocked time.Time `json:"last_restocked"`
	Status        string    `json:"status"` // ok, low, out
	Price         float64   `json:"price"`
}

// GetInventory retrieves all inventory items with current stock levels
func (h *InventoryHandler) GetInventory(c *gin.Context) {
	query := `
		SELECT 
			p.id, p.name, c.name as category_name,
			COALESCE(i.current_stock, 0) as current_stock,
			COALESCE(i.minimum_stock, 10) as min_stock,
			COALESCE(i.maximum_stock, 100) as max_stock,
			'pcs' as unit,
			COALESCE(i.last_restocked_at, p.created_at) as last_restocked,
			p.price,
			CASE 
				WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
				WHEN COALESCE(i.current_stock, 0) < COALESCE(i.minimum_stock, 10) THEN 'low'
				ELSE 'ok'
			END as status
		FROM products p
		LEFT JOIN categories c ON p.category_id = c.id
		LEFT JOIN inventory i ON p.id = i.product_id
		WHERE p.is_available = true
		ORDER BY status DESC, c.name, p.name
	`

	rows, err := h.db.Query(query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch inventory"})
		return
	}
	defer rows.Close()

	var items []InventoryItem
	for rows.Next() {
		var item InventoryItem
		err := rows.Scan(
			&item.ProductID, &item.ProductName, &item.CategoryName,
			&item.CurrentStock, &item.MinStock, &item.MaxStock,
			&item.Unit, &item.LastRestocked, &item.Price, &item.Status,
		)
		if err != nil {
			continue
		}
		items = append(items, item)
	}

	if items == nil {
		items = []InventoryItem{}
	}

	c.JSON(http.StatusOK, items)
}

// GetProductInventory retrieves inventory for a specific product
func (h *InventoryHandler) GetProductInventory(c *gin.Context) {
	productID := c.Param("product_id")

	var item InventoryItem
	query := `
		SELECT 
			p.id, p.name, c.name as category_name,
			COALESCE(i.current_stock, 0) as current_stock,
			COALESCE(i.min_stock, 10) as min_stock,
			COALESCE(i.max_stock, 100) as max_stock,
			COALESCE(i.unit, 'pcs') as unit,
			COALESCE(i.last_restocked, p.created_at) as last_restocked,
			p.price,
			CASE 
				WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
				WHEN COALESCE(i.current_stock, 0) < COALESCE(i.min_stock, 10) THEN 'low'
				ELSE 'ok'
			END as status
		FROM products p
		LEFT JOIN categories c ON p.category_id = c.id
		LEFT JOIN inventory i ON p.id = i.product_id
		WHERE p.id = $1
	`

	err := h.db.QueryRow(query, productID).Scan(
		&item.ProductID, &item.ProductName, &item.CategoryName,
		&item.CurrentStock, &item.MinStock, &item.MaxStock,
		&item.Unit, &item.LastRestocked, &item.Price, &item.Status,
	)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Product not found"})
		return
	}

	c.JSON(http.StatusOK, item)
}

// AdjustStockRequest represents a stock adjustment request
type AdjustStockRequest struct {
	ProductID uuid.UUID `json:"product_id" binding:"required"`
	Operation string    `json:"operation" binding:"required"` // "add" or "remove"
	Quantity  int       `json:"quantity" binding:"required,gt=0"`
	Reason    string    `json:"reason" binding:"required"`
	Notes     string    `json:"notes"`
}

// AdjustStock adjusts inventory stock levels with audit trail
func (h *InventoryHandler) AdjustStock(c *gin.Context) {
	var req AdjustStockRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Validate operation
	if req.Operation != "add" && req.Operation != "remove" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Operation must be 'add' or 'remove'"})
		return
	}

	// Validate reason
	validReasons := []string{"purchase", "sale", "spoilage", "manual_adjustment", "inventory_count", "return", "damage", "theft", "expired"}
	validReason := false
	for _, r := range validReasons {
		if req.Reason == r {
			validReason = true
			break
		}
	}
	if !validReason {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid reason"})
		return
	}

	// Get user ID from context
	userID, exists := c.Get("user_id")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}

	// Begin transaction
	tx, err := h.db.Begin()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to start transaction"})
		return
	}
	defer tx.Rollback()

	// Get or create inventory record
	var currentStock int
	var inventoryID uuid.UUID
	checkQuery := `SELECT id, current_stock FROM inventory WHERE product_id = $1`
	err = tx.QueryRow(checkQuery, req.ProductID).Scan(&inventoryID, &currentStock)

	if err == sql.ErrNoRows {
		// Create new inventory record
		inventoryID = uuid.New()
		currentStock = 0
		insertQuery := `
			INSERT INTO inventory (id, product_id, current_stock, minimum_stock, maximum_stock)
			VALUES ($1, $2, 0, 10, 100)
		`
		_, err = tx.Exec(insertQuery, inventoryID, req.ProductID)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create inventory record"})
			return
		}
	} else if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch inventory"})
		return
	}

	// Calculate new stock
	previousStock := currentStock
	var newStock int
	if req.Operation == "add" {
		newStock = currentStock + req.Quantity
	} else {
		newStock = currentStock - req.Quantity
		if newStock < 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Insufficient stock"})
			return
		}
	}

	// Update inventory
	updateQuery := `
		UPDATE inventory 
		SET current_stock = $1, last_restocked_at = NOW(), updated_at = NOW()
		WHERE product_id = $2
	`
	_, err = tx.Exec(updateQuery, newStock, req.ProductID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update inventory"})
		return
	}

	// Create history record
	historyQuery := `
		INSERT INTO inventory_history (product_id, operation, quantity, previous_stock, new_stock, reason, notes, adjusted_by)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
	`
	_, err = tx.Exec(historyQuery, req.ProductID, req.Operation, req.Quantity, previousStock, newStock, req.Reason, req.Notes, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create history record"})
		return
	}

	// Commit transaction
	if err = tx.Commit(); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to commit transaction"})
		return
	}

	// Check for low stock and send notification
	var productName string
	var minStock int
	stockQuery := `
		SELECT p.name, COALESCE(i.minimum_stock, 10) 
		FROM products p 
		LEFT JOIN inventory i ON p.id = i.product_id 
		WHERE p.id = $1
	`
	h.db.QueryRow(stockQuery, req.ProductID).Scan(&productName, &minStock)

	if newStock < minStock {
		// Send low stock notification (ignore errors)
		go h.notificationSvc.NotifyLowStock(productName, newStock, minStock)
	}

	c.JSON(http.StatusOK, gin.H{
		"message":        "Stock adjusted successfully",
		"previous_stock": previousStock,
		"new_stock":      newStock,
	})
}

// GetLowStock retrieves items below minimum stock threshold
func (h *InventoryHandler) GetLowStock(c *gin.Context) {
	query := `
		SELECT 
			p.id, p.name, c.name as category_name,
			COALESCE(i.current_stock, 0) as current_stock,
			COALESCE(i.minimum_stock, 10) as min_stock,
			COALESCE(i.maximum_stock, 100) as max_stock,
			'pcs' as unit,
			COALESCE(i.last_restocked_at, p.created_at) as last_restocked,
			p.price,
			CASE 
				WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
				ELSE 'low'
			END as status
		FROM products p
		LEFT JOIN categories c ON p.category_id = c.id
		LEFT JOIN inventory i ON p.id = i.product_id
		WHERE p.is_available = true 
		  AND COALESCE(i.current_stock, 0) < COALESCE(i.minimum_stock, 10)
		ORDER BY COALESCE(i.current_stock, 0) ASC, p.name
	`

	rows, err := h.db.Query(query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch low stock items"})
		return
	}
	defer rows.Close()

	var items []InventoryItem
	for rows.Next() {
		var item InventoryItem
		err := rows.Scan(
			&item.ProductID, &item.ProductName, &item.CategoryName,
			&item.CurrentStock, &item.MinStock, &item.MaxStock,
			&item.Unit, &item.LastRestocked, &item.Price, &item.Status,
		)
		if err != nil {
			continue
		}
		items = append(items, item)
	}

	if items == nil {
		items = []InventoryItem{}
	}

	c.JSON(http.StatusOK, items)
}

// HistoryRecord represents a stock movement history entry
type HistoryRecord struct {
	ID            uuid.UUID `json:"id"`
	Operation     string    `json:"operation"`
	Quantity      int       `json:"quantity"`
	PreviousStock int       `json:"previous_stock"`
	NewStock      int       `json:"new_stock"`
	Reason        string    `json:"reason"`
	Notes         string    `json:"notes"`
	AdjustedBy    string    `json:"adjusted_by"` // username
	CreatedAt     time.Time `json:"created_at"`
}

// GetStockHistory retrieves stock movement history for a product
func (h *InventoryHandler) GetStockHistory(c *gin.Context) {
	productID := c.Param("product_id")

	query := `
		SELECT 
			ih.id, ih.operation, ih.quantity, ih.previous_stock, ih.new_stock,
			ih.reason, COALESCE(ih.notes, '') as notes, 
			COALESCE(u.username, 'System') as adjusted_by,
			ih.created_at
		FROM inventory_history ih
		LEFT JOIN users u ON ih.adjusted_by = u.id
		WHERE ih.product_id = $1
		ORDER BY ih.created_at DESC
		LIMIT 100
	`

	rows, err := h.db.Query(query, productID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch history"})
		return
	}
	defer rows.Close()

	var history []HistoryRecord
	for rows.Next() {
		var record HistoryRecord
		err := rows.Scan(
			&record.ID, &record.Operation, &record.Quantity,
			&record.PreviousStock, &record.NewStock, &record.Reason,
			&record.Notes, &record.AdjustedBy, &record.CreatedAt,
		)
		if err != nil {
			continue
		}
		history = append(history, record)
	}

	if history == nil {
		history = []HistoryRecord{}
	}

	c.JSON(http.StatusOK, history)
}

File: backend/src/handlers/inventory.ts
import type { Context } from 'hono';
import { sql } from 'drizzle-orm';
import { db, pool } from '../db/connection.js';

// ── GetInventory ──────────────────────────────────────────────────────────

export async function getInventory(c: Context) {
  try {
    const rows = await db.execute<{
      product_id: string;
      product_name: string;
      category_name: string;
      current_stock: number;
      min_stock: number;
      max_stock: number;
      unit: string;
      last_restocked: string;
      price: string;
      status: string;
    }>(sql`
      SELECT
        p.id as product_id, p.name as product_name, c.name as category_name,
        COALESCE(i.current_stock, 0) as current_stock,
        COALESCE(i.minimum_stock, 10) as min_stock,
        COALESCE(i.maximum_stock, 100) as max_stock,
        'pcs' as unit,
        COALESCE(i.last_restocked_at, p.created_at) as last_restocked,
        p.price,
        CASE
          WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
          WHEN COALESCE(i.current_stock, 0) < COALESCE(i.minimum_stock, 10) THEN 'low'
          ELSE 'ok'
        END as status
      FROM products p
      LEFT JOIN categories c ON p.category_id = c.id
      LEFT JOIN inventory i ON p.id = i.product_id
      WHERE p.is_available = true
      ORDER BY status DESC, c.name, p.name
    `);

    const items = rows.rows.map((row) => ({
      product_id: row.product_id,
      product_name: row.product_name,
      category_name: row.category_name,
      current_stock: Number(row.current_stock),
      min_stock: Number(row.min_stock),
      max_stock: Number(row.max_stock),
      unit: row.unit,
      last_restocked: row.last_restocked,
      price: Number(row.price),
      status: row.status,
    }));

    // Return raw array (matches Go behavior)
    return c.json(items, 200);
  } catch {
    return c.json({ error: 'Failed to fetch inventory' }, 500);
  }
}

// ── GetProductInventory ──────────────────────────────────────────────────────

export async function getProductInventory(c: Context) {
  const productId = c.req.param('product_id');

  try {
    const rows = await db.execute<{
      product_id: string;
      product_name: string;
      category_name: string;
      current_stock: number;
      min_stock: number;
      max_stock: number;
      unit: string;
      last_restocked: string;
      price: string;
      status: string;
    }>(sql`
      SELECT
        p.id as product_id, p.name as product_name, c.name as category_name,
        COALESCE(i.current_stock, 0) as current_stock,
        COALESCE(i.minimum_stock, 10) as min_stock,
        COALESCE(i.maximum_stock, 100) as max_stock,
        COALESCE(i.unit_cost, '0') as unit,
        COALESCE(i.last_restocked_at, p.created_at) as last_restocked,
        p.price,
        CASE
          WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
          WHEN COALESCE(i.current_stock, 0) < COALESCE(i.minimum_stock, 10) THEN 'low'
          ELSE 'ok'
        END as status
      FROM products p
      LEFT JOIN categories c ON p.category_id = c.id
      LEFT JOIN inventory i ON p.id = i.product_id
      WHERE p.id = ${productId}
    `);

    if (rows.rows.length === 0) {
      return c.json({ error: 'Product not found' }, 404);
    }

    const row = rows.rows[0];
    return c.json({
      product_id: row.product_id,
      product_name: row.product_name,
      category_name: row.category_name,
      current_stock: Number(row.current_stock),
      min_stock: Number(row.min_stock),
      max_stock: Number(row.max_stock),
      unit: 'pcs',
      last_restocked: row.last_restocked,
      price: Number(row.price),
      status: row.status,
    }, 200);
  } catch {
    return c.json({ error: 'Product not found' }, 404);
  }
}

// ── AdjustStock ──────────────────────────────────────────────────────────

export async function adjustStock(c: Context) {
  let body: {
    product_id: string;
    operation: string;
    quantity: number;
    reason: string;
    notes?: string;
  };

  try {
    body = await c.req.json();
  } catch {
    return c.json({ error: 'Invalid request body' }, 400);
  }

  if (!body.product_id) {
    return c.json({ error: 'product_id is required' }, 400);
  }
  if (!body.operation) {
    return c.json({ error: 'operation is required' }, 400);
  }
  if (!body.quantity || body.quantity <= 0) {
    return c.json({ error: 'quantity must be greater than 0' }, 400);
  }
  if (!body.reason) {
    return c.json({ error: 'reason is required' }, 400);
  }

  // Validate operation
  if (body.operation !== 'add' && body.operation !== 'remove') {
    return c.json({ error: "Operation must be 'add' or 'remove'" }, 400);
  }

  // Validate reason
  const validReasons = ['purchase', 'sale', 'spoilage', 'manual_adjustment', 'inventory_count', 'return', 'damage', 'theft', 'expired'];
  if (!validReasons.includes(body.reason)) {
    return c.json({ error: 'Invalid reason' }, 400);
  }

  const userId = c.get('user_id');

  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    // Get or create inventory record
    let currentStock = 0;
    const checkRes = await client.query(
      'SELECT id, current_stock FROM inventory WHERE product_id = $1',
      [body.product_id],
    );

    if (checkRes.rows.length === 0) {
      // Create new inventory record
      await client.query(
        'INSERT INTO inventory (product_id, current_stock, minimum_stock, maximum_stock) VALUES ($1, 0, 10, 100)',
        [body.product_id],
      );
    } else {
      currentStock = Number(checkRes.rows[0].current_stock);
    }

    // Calculate new stock
    const previousStock = currentStock;
    let newStock: number;
    if (body.operation === 'add') {
      newStock = currentStock + body.quantity;
    } else {
      newStock = currentStock - body.quantity;
      if (newStock < 0) {
        await client.query('ROLLBACK');
        return c.json({ error: 'Insufficient stock' }, 400);
      }
    }

    // Update inventory
    await client.query(
      'UPDATE inventory SET current_stock = $1, last_restocked_at = NOW(), updated_at = NOW() WHERE product_id = $2',
      [newStock, body.product_id],
    );

    // Create history record
    await client.query(
      `INSERT INTO inventory_history (product_id, operation, quantity, previous_stock, new_stock, reason, notes, adjusted_by)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
      [body.product_id, body.operation, body.quantity, previousStock, newStock, body.reason, body.notes || null, userId],
    );

    await client.query('COMMIT');

    return c.json({
      message: 'Stock adjusted successfully',
      previous_stock: previousStock,
      new_stock: newStock,
    }, 200);
  } catch {
    await client.query('ROLLBACK');
    return c.json({ error: 'Failed to adjust stock' }, 500);
  } finally {
    client.release();
  }
}

// ── GetLowStock ──────────────────────────────────────────────────────────

export async function getLowStock(c: Context) {
  try {
    const rows = await db.execute<{
      product_id: string;
      product_name: string;
      category_name: string;
      current_stock: number;
      min_stock: number;
      max_stock: number;
      unit: string;
      last_restocked: string;
      price: string;
      status: string;
    }>(sql`
      SELECT
        p.id as product_id, p.name as product_name, c.name as category_name,
        COALESCE(i.current_stock, 0) as current_stock,
        COALESCE(i.minimum_stock, 10) as min_stock,
        COALESCE(i.maximum_stock, 100) as max_stock,
        'pcs' as unit,
        COALESCE(i.last_restocked_at, p.created_at) as last_restocked,
        p.price,
        CASE
          WHEN COALESCE(i.current_stock, 0) = 0 THEN 'out'
          ELSE 'low'
        END as status
      FROM products p
      LEFT JOIN categories c ON p.category_id = c.id
      LEFT JOIN inventory i ON p.id = i.product_id
      WHERE p.is_available = true
        AND COALESCE(i.current_stock, 0) < COALESCE(i.minimum_stock, 10)
      ORDER BY COALESCE(i.current_stock, 0) ASC, p.name
    `);

    const items = rows.rows.map((row) => ({
      product_id: row.product_id,
      product_name: row.product_name,
      category_name: row.category_name,
      current_stock: Number(row.current_stock),
      min_stock: Number(row.min_stock),
      max_stock: Number(row.max_stock),
      unit: row.unit,
      last_restocked: row.last_restocked,
      price: Number(row.price),
      status: row.status,
    }));

    // Return raw array (matches Go behavior)
    return c.json(items, 200);
  } catch {
    return c.json({ error: 'Failed to fetch low stock items' }, 500);
  }
}

// ── GetStockHistory ──────────────────────────────────────────────────────────

export async function getStockHistory(c: Context) {
  const productId = c.req.param('product_id');

  try {
    const rows = await db.execute<{
      id: string;
      operation: string;
      quantity: number;
      previous_stock: number;
      new_stock: number;
      reason: string;
      notes: string;
      adjusted_by: string;
      created_at: string;
    }>(sql`
      SELECT
        ih.id, ih.operation, ih.quantity, ih.previous_stock, ih.new_stock,
        ih.reason, COALESCE(ih.notes, '') as notes,
        COALESCE(u.username, 'System') as adjusted_by,
        ih.created_at
      FROM inventory_history ih
      LEFT JOIN users u ON ih.adjusted_by = u.id
      WHERE ih.product_id = ${productId}
      ORDER BY ih.created_at DESC
      LIMIT 100
    `);

    const history = rows.rows.map((row) => ({
      id: row.id,
      operation: row.operation,
      quantity: Number(row.quantity),
      previous_stock: Number(row.previous_stock),
      new_stock: Number(row.new_stock),
      reason: row.reason,
      notes: row.notes,
      adjusted_by: row.adjusted_by,
      created_at: row.created_at,
    }));

    // Return raw array (matches Go behavior)
    return c.json(history, 200);
  } catch {
    return c.json({ error: 'Failed to fetch history' }, 500);
  }
}
